### PR TITLE
8292541: [Metrics] Reported memory limit may exceed physical machine memory

### PR DESCRIPTION
**Pull request by jmtd**
When the container memory exceeds the physical host's memory, the Java metrics reporting is wrong.

https://bugs.openjdk.org/browse/JDK-8292541

This is the equivalent core-libs fix for Hotspot's [JDK-8292083](https://bugs.openjdk.org/browse/JDK-8292083)
(https://github.com/openjdk/jdk/pull/9880)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292541](https://bugs.openjdk.org/browse/JDK-8292541): [Metrics] Reported memory limit may exceed physical machine memory

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [ff6990cb](https://git.openjdk.org/jdk/pull/10017/files/ff6990cb3ec235fe21f90c812d8816de368e2f23)
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10017/head:pull/10017` \
`$ git checkout pull/10017`

Update a local copy of the PR: \
`$ git checkout pull/10017` \
`$ git pull https://git.openjdk.org/jdk pull/10017/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10017`

View PR using the GUI difftool: \
`$ git pr show -t 10017`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10017.diff">https://git.openjdk.org/jdk/pull/10017.diff</a>

</details>

### CODE REVIEW

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (tstuefe) [inline src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java:126]: nit: Start sentences uppercase

File: src/java.base/linux/native/libjava/CgroupMetrics.c
Comment (tstuefe) [inline src/java.base/linux/native/libjava/CgroupMetrics.c:41]: Why not do it the same way hotspot does? 
```
  sysinfo(&si);
  avail_mem = (julong)si.freeram * si.mem_unit;
```
if for some weird reason the APIs return different numbers, at least we use the same numbers in JDK and VM.

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (jmtd) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java:126]: Thanks - Address in ff6990cb3ec

File: src/java.base/linux/native/libjava/CgroupMetrics.c
Comment (jmtd) [reply src/java.base/linux/native/libjava/CgroupMetrics.c:41]: I don't have a strong preference for `sysconf` over `sysinfo`: The former avoids an explicit local variable but I guess that doesn't really matter. I take your point about possible divergent values from hotspot. Unfortunately there's a third example, `Java_com_sun_management_internal_OperatingSystemImpl_getFreeMemorySize0` in [src/jdk.management/unix/native/libmanagement_ext/OperatingSystemImpl.c](https://github.com/openjdk/jdk/blob/master/src/jdk.management/unix/native/libmanagement_ext/OperatingSystemImpl.c#L249) which is where I took the `sysconf` approach from.

(I don't think it is practical for the code I've changed here, `CgroupMetrics`, to re-use the existing (currently private) native method `getTotalMemorySize0` in `OperatingSystemImpl` due to circular dependency issues).

With respect to consistency with hotspot: out of scope for these two bugs, perhaps hotspot could export the data structures it populates from cgroups for internal (to the jdk) consumption, and we could avoid re-parsing the cgroup structures in Java code entirely. That would guarantee consistency and mean that the consumers benefit from e.g. the caching that the hotspot code has and this lacks. This might make a good RFE?

Comment (tstuefe) [reply src/java.base/linux/native/libjava/CgroupMetrics.c:41]: Okay, lets leave it this way. One would hope the results are identical.

> With respect to consistency with hotspot: out of scope for these two bugs, perhaps hotspot could export the data structures it populates from cgroups for internal (to the jdk) consumption, and we could avoid re-parsing the cgroup structures in Java code entirely. That would guarantee consistency and mean that the consumers benefit from e.g. the caching that the hotspot code has and this lacks. This might make a good RFE?

Possibly. Personally, I would not broaden the JVM interface for that, but I have not looked deeply enough into the java implementation to weigh in.

**Review by tstuefe — APPROVED** [review]
LGTM

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java:124]: Style nit: `subsMem` camelcase ;-)

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (jmtd) [inline test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:53]: I've just realised this will not match (yet) since the relevant log line is added in #9880 .

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (jmtd) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java]: Addressed in 01d07ab4b2e.

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (jmtd) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java]: Addressed in 01d07ab4b2e.

File: src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java
Comment (jmtd) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupMetrics.java]: Addressed in 85fa402acbb.

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (jerboaa) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:53]: Yes. There is a feature to use dependent PRs, which could be used here. Just target `pr/9880` instead of `master`.

**Review by jerboaa — APPROVED** [review]
Looks good.

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (jmtd) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:53]: I've pushed a small excerpt of the changes from #9880 in 1bf3aa341fd93079289f11e620938e6bed40b382 to this PR in order to add the log lines that the tests depend on.

Comment (jerboaa) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:53]: We cannot have this in both PRs, though.

Comment (jmtd) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:53]: Indeed: even without this addition, the other one will merge conflict with this PR (assuming this one is merged first). So I can deal with the duplication when I resolve the conflict.

**Review by jerboaa — CHANGES_REQUESTED** [review]
Sorry, but this last change is a hack.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jmtd) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp:532]: Fair enough. You're right, it only works because the first match is prior to cgroups overwriting the physical number. I'm happy to wait until we merge #9880.

**Comment by jmtd** [comment]
Now that #9880 has been merged the hack is gone and the tests can match against the log lines to establish base line physical RAM. There are three tests that use this now. In [cc8fdf3](https://github.com/openjdk/jdk/pull/10017/commits/cc8fdf36af5e3f75c47348d25697e8a0b496dd42) for this PR I've adjusted the first test (merged already) to share the established baseline physical RAM value rather than calculate it again.

@jerboaa WDYT now? And thanks for all your support!

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:93]: Could we pass `hostMaxMem` to the relevant test functions as a parameter, instead of relying on a static variable magically be set, please? It would be a good symmetry to other tests.

Comment (jmtd) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java]: Good idea! Implemented in 0f37be83ff0.

**Review by jerboaa — APPROVED** [review]
OK.

**Pull request closed without merging**
